npm run test : 'setImmediate' is defined but never used

When I try to run the tests, I get this error:

```
> jshint mqtt.js lib/* test/*.js test/browser/*.js test/helpers/*.js

lib/client.js: line 5, col 1, 'setImmediate' is defined but never used.

test/abstract_client.js: line 2, col 1, 'setImmediate' is defined but never used.

test/client.js: line 2, col 1, 'setImmediate' is defined but never used.

test/secure_client.js: line 2, col 1, 'setImmediate' is defined but never used.

test/util.js: line 2, col 1, 'setImmediate' is defined but never used.

test/websocket_client.js: line 5, col 1, 'setImmediate' is defined but never used.

6 errors

npm ERR! mqtt@1.5.0 check-style:jshint: `jshint mqtt.js lib/* test/*.js test/browser/*.js test/helpers/*.js`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the mqtt@1.5.0 check-style:jshint script.
npm ERR! This is most likely a problem with the mqtt package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     jshint mqtt.js lib/* test/*.js test/browser/*.js test/helpers/*.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls mqtt
npm ERR! There is likely additional logging output above.
npm ERR! System Linux 3.16.0-30-generic
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "run" "check-style:jshint"
npm ERR! cwd /home/arzael/MQTT/MQTT.js
npm ERR! node -v v0.10.37
npm ERR! npm -v 1.4.28
npm ERR! code ELIFECYCLE
```

What is `setImmediate` supposed to be used for ? Should it be deleted ?

(It is in the last version of MQTT.js)

--- a/lib/client.js
+++ b/lib/client.js
@@ -554,6 +554,7 @@ MqttClient.prototype._clearReconnect = function () {
  * @api private
  */
 MqttClient.prototype._cleanUp = function (forced, done) {
+
   if (done) {
     this.stream.on('close', done);
   }
@@ -570,7 +571,10 @@ MqttClient.prototype._cleanUp = function (forced, done) {
     );
   }
 
-  this._clearReconnect();
+  if (this.reconnectTimer) {
+    this._clearReconnect();
+    this._setupReconnect();
+  }
 
   if (null !== this.pingTimer) {
     this.pingTimer.clear();

--- a/lib/client.js
+++ b/lib/client.js
@@ -9,6 +9,7 @@ var events = require('events'),
   mqttPacket = require('mqtt-packet'),
   Writable = require('readable-stream').Writable,
   inherits = require('inherits'),
+  reInterval = require('reinterval'),
   setImmediate = global.setImmediate || function (callback) {
     // works in node v0.8
     process.nextTick(callback);
@@ -176,7 +177,7 @@ function MqttClient (streamBuilder, options) {
   // Clear ping timer
   this.on('close', function () {
     if (null !== that.pingTimer) {
-      clearInterval(that.pingTimer);
+      that.pingTimer.clear();
       that.pingTimer = null;
     }
   });
@@ -284,6 +285,8 @@ MqttClient.prototype._handlePacket = function (packet, done) {
       // or just log it
       break;
   }
+  // When a packet is received, reschedule the ping timer
+  this._shiftPingInterval();
 };
 
 MqttClient.prototype._checkDisconnecting = function (callback) {
@@ -570,7 +573,7 @@ MqttClient.prototype._cleanUp = function (forced, done) {
   this._clearReconnect();
 
   if (null !== this.pingTimer) {
-    clearInterval(this.pingTimer);
+    this.pingTimer.clear();
     this.pingTimer = null;
   }
 };
@@ -615,12 +618,22 @@ MqttClient.prototype._setupPingTimer = function () {
 
   if (!this.pingTimer && this.options.keepalive) {
     this.pingResp = true;
-    this.pingTimer = setInterval(function () {
+    this.pingTimer = reInterval(function () {
       that._checkPing();
     }, this.options.keepalive * 1000);
   }
 };
 
+/**
+ * _shiftPingInterval - reschedule the ping interval
+ *
+ * @api private
+ */
+MqttClient.prototype._shiftPingInterval = function () {
+  if (this.pingTimer && this.options.keepalive) {
+    this.pingTimer.reschedule(this.options.keepalive * 1000);
+  }
+};
 /**
  * _checkPing - check if a pingresp has come back, and ping the server again
  *

--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "mqtt",
   "description": "A library for the MQTT protocol",
-  "version": "1.5.1",
+  "version": "1.5.0",
   "contributors": [
     "Adam Rudd <adamvrr@gmail.com>",
     "Matteo Collina <matteo.collina@gmail.com> (https://github.com/mcollina)"

--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "mqtt",
   "description": "A library for the MQTT protocol",
-  "version": "1.5.0",
+  "version": "1.5.1",
   "contributors": [
     "Adam Rudd <adamvrr@gmail.com>",
     "Matteo Collina <matteo.collina@gmail.com> (https://github.com/mcollina)"

--- a/package.json
+++ b/package.json
@@ -62,7 +62,7 @@
     "browserify": "^11.0.1",
     "eslint": "^1.3.1",
     "jscs": "^2.1.1",
-    "jshint": "^2.6.2",
+    "jshint": "2.7.0",
     "mocha": "*",
     "pre-commit": "1.1.1",
     "should": "*",

--- a/package.json
+++ b/package.json
@@ -54,6 +54,7 @@
     "mqtt-connection": "^2.0.0",
     "mqtt-packet": "^3.2.0",
     "readable-stream": "~1.0.2",
+    "reinterval": "^1.0.1",
     "websocket-stream": "^2.0.2",
     "xtend": "^4.0.0"
   },

--- a/test/abstract_client.js
+++ b/test/abstract_client.js
@@ -608,7 +608,6 @@ module.exports = function (server, config) {
   });
 
   describe('pinging', function () {
-
     it('should set a ping timer', function (done) {
       var client = connect({keepalive: 3});
       client.once('connect', function () {
@@ -645,19 +644,39 @@ module.exports = function (server, config) {
       });
       setTimeout(done, 1000);
     });
-    it('should defer the next ping when a control packer is received', function (done) {
-      var client = connect({keepalive: 100});
-      client.subscribe('test');
-      client._checkPing = sinon.spy();
+    it('should defer the next ping when a control packet is received', function (done) {
+      var client = connect({keepalive: 0.1});
 
-      setTimeout(function () {
-        client.publish('test');
-      }, 75);
+      var testPacket = {
+          topic: 'test',
+          payload: 'message',
+          retain: true,
+          qos: 1,
+          messageId: 42
+        };
 
-      setTimeout(function () {
-        client._checkPing.callCount.should.equal(0);
-        done();
-      }, 125);
+      server.once('client', function (serverClient) {
+        client.subscribe('test', function () {
+
+          serverClient.publish(testPacket);
+          client._checkPing = sinon.spy();
+
+          setTimeout(function () {
+            client._checkPing.callCount.should.equal(0);
+            serverClient.publish(testPacket);
+
+            setTimeout(function () {
+              client._checkPing.callCount.should.equal(0);
+              serverClient.publish(testPacket);
+
+              setTimeout(function () {
+                client._checkPing.callCount.should.equal(0);
+                done();
+              }, 75);
+            }, 75);
+          }, 75);
+        });
+      });
     });
   });
 

--- a/test/abstract_client.js
+++ b/test/abstract_client.js
@@ -645,9 +645,9 @@ module.exports = function (server, config) {
       setTimeout(done, 1000);
     });
     it('should defer the next ping when a control packet is received', function (done) {
-      var client = connect({keepalive: 0.1});
+      var client = connect({keepalive: 0.1}),
 
-      var testPacket = {
+        testPacket = {
           topic: 'test',
           payload: 'message',
           retain: true,

--- a/test/abstract_client.js
+++ b/test/abstract_client.js
@@ -645,6 +645,20 @@ module.exports = function (server, config) {
       });
       setTimeout(done, 1000);
     });
+    it('should defer the next ping when a control packer is received', function (done) {
+      var client = connect({keepalive: 100});
+      client.subscribe('test');
+      client._checkPing = sinon.spy();
+
+      setTimeout(function () {
+        client.publish('test');
+      }, 75);
+
+      setTimeout(function () {
+        client._checkPing.callCount.should.equal(0);
+        done();
+      }, 125);
+    });
   });
 
   describe('subscribing', function () {

